File: 05/soln_2.py
class Line:
    def __init__(self, x1, y1, x2, y2):
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2


def sgn(x):
    return 1 if x > 0 else (-1 if x < 0 else 0)


class Grid:
    def __init__(self, width, height):
        self.grid = [[0 for _ in range(height)] for _ in range(width)]

    def cover(self, line):
        x1, x2, y1, y2 = line.x1, line.x2, line.y1, line.y2
        x_step = sgn(x2 - x1)
        y_step = sgn(y2 - y1)
        while x1 != x2 + x_step or y1 != y2 + y_step:
            self.grid[x1][y1] += 1
            x1 += x_step
            y1 += y_step

    def count_overlaps(self):
        return sum(covers > 1 for row in self.grid for covers in row)


def parse_line(s: str) -> Line:
    start, end = s.split("->")
    x1, y1 = map(int, start.split(","))
    x2, y2 = map(int, end.split(","))
    line = Line(x1, y1, x2, y2)
    return line


def main():
    with open("input.txt") as f:
        inp = f.readlines()
    lines = map(parse_line, inp)
    grid = Grid(1001, 1001)
    for line in lines:
        grid.cover(line)
    print(grid.count_overlaps())


if __name__ == "__main__":
    main()

File: 05/soln_2.exs
defmodule Soln do
  def sgn(x) when x > 0, do: 1
  def sgn(x) when x == 0, do: 0
  def sgn(x) when x < 0, do: -1

  def parse_coords(s) do
    String.split(s, ",")
    |> Enum.map(&String.trim/1)
    |> Enum.map(
      &(Integer.parse(&1)
        |> elem(0))
    )
  end

  def handle_line(s) do
    # return a map where the keys represent the points covered by the given line
    [start, end_] = String.split(s, "->")
    [x1, y1] = parse_coords(start)
    [x2, y2] = parse_coords(end_)
    xd = x2 - x1
    yd = y2 - y1
    x_step = sgn(xd)
    y_step = sgn(yd)
    steps = max(x_step * xd, y_step * yd)

    Enum.reduce(0..steps, %{}, fn step, acc ->
      Map.put(acc, {x1 + x_step * step, y1 + y_step * step}, 1)
    end)
  end
end

# pointless concurrent processing
File.stream!("input.txt")
|> Task.async_stream(&Soln.handle_line/1)
|> Enum.map(&elem(&1, 1))
|> Enum.reduce(fn x, y -> Map.merge(x, y, fn _, v1, v2 -> v1 + v2 end) end)
|> Enum.count(fn {_, cover_count} -> cover_count > 1 end)
|> IO.puts()
